exonum/src/explorer/mod.rs

```diff
@@ -0,0 +1,976 @@
+                "Block not found, height: {:?}",
+                height
+            ));
+            blocks.get(&block_hash).expect(&format!(
+                "Block not found, hash: {:?}",
+                block_hash
+            ))
+        };
+
+        BlockInfo {
+            explorer,
+            block,
+            precommits: RefCell::new(None),
+            txs: RefCell::new(None),
+        }
+    }
+
+    /// Returns the block header as recorded in the blockchain.
+    pub fn block(&self) -> &Block {
+        &self.block
+    }
+
+    /// Returns the number of transactions in this block.
+    pub fn len(&self) -> usize {
+        self.block.tx_count() as usize
+    }
+
+    /// Is this block empty (i.e., contains no transactions)?
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Returns a list of precommits for this block.
+    pub fn precommits(&self) -> Ref<[Precommit]> {
+        if self.precommits.borrow().is_none() {
+            let precommits = self.explorer.precommits(&self.block);
+            *self.precommits.borrow_mut() = Some(precommits);
+        }
+
+        Ref::map(self.precommits.borrow(), |cache| {
+            cache.as_ref().unwrap().as_ref()
+        })
+    }
+
+    /// List of hashes for transactions that was executed into this block.
+    pub fn transaction_hashes(&self) -> Ref<[Hash]> {
+        if self.txs.borrow().is_none() {
+            let txs = self.explorer.transaction_hashes(&self.block);
+            *self.txs.borrow_mut() = Some(txs);
+        }
+
+        Ref::map(self.txs.borrow(), |cache| cache.as_ref().unwrap().as_ref())
+    }
+
+    /// Returns a transaction with the specified index in the block.
+    pub fn transaction(&self, index: usize) -> Option<CommittedTransaction> {
+        self.transaction_hashes().get(index).map(|hash| {
+            self.explorer.committed_transaction(hash, None)
+        })
+    }
+
+    /// Iterates over transactions in the block.
+    pub fn iter(&self) -> Transactions {
+        Transactions {
+            block: self,
+            ptr: 0,
+            len: self.len(),
+        }
+    }
+
+    /// Loads transactions and precommits for the block.
+    pub fn with_transactions(self) -> BlockWithTransactions {
+        let (explorer, block, precommits, transactions) =
+            (self.explorer, self.block, self.precommits, self.txs);
+
+        let precommits = precommits.into_inner().unwrap_or_else(
+            || explorer.precommits(&block),
+        );
+        let transactions = transactions
+            .into_inner()
+            .unwrap_or_else(|| explorer.transaction_hashes(&block))
+            .iter()
+            .map(|tx_hash| explorer.committed_transaction(tx_hash, None))
+            .collect();
+
+        BlockWithTransactions {
+            block,
+            precommits,
+            transactions,
+        }
+    }
+}
+
+impl<'a> Deref for BlockInfo<'a> {
```

Comment:
Do `Deref` implementations for `BlockInfo` and `BlockWithTransactions` make sense?

Reply:
I think it is a bad idea to use `Deref` trait to emulate polymorphism. To quote [the documentation](https://doc.rust-lang.org/std/ops/trait.Deref.html):

> Because of this, Deref should only be implemented for smart pointers to avoid confusion.

 It might be confusing for rust beginners and it generally feels like a hack. We are supposed to embrace Rust language design choices (like absence of inheritance) not to hack around it with `Deref`.

I think that a simple `block(&self)` getter is enough, maybe `block_mut(&mut self)` and `into_block(self)` too if necessary. This is how it is done all over the standard library and is more familiar to rust developers.

Reply:
Got it; I had similar concerns initially. Will rework to use `block(&self)` and `into_block(self)` (or maybe `*header`; imo, it would be a better naming choice).